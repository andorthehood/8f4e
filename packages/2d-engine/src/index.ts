// Export types
export type { SpriteCoordinates, SpriteLookup } from './types';
export type { PostProcessEffect, PostProcessPipeline, EffectUniforms, UniformBufferMapping } from './types/postProcess';

// Export main Engine class (public API)
export { Engine } from './engine';

// Export Renderer for advanced users who want low-level access
export { Renderer } from './renderer';

// Export post-processing system
export { PostProcessManager } from './postProcess/PostProcessManager';
