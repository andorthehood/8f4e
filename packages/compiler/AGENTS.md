# Repository Guidelines

## Package Scope & Layout
- Path: `packages/compiler`; source in `src/`, output in `dist/` (ESM).
- Consumed via alias `@8f4e/compiler`.

## Build, Test, Dev
- From root: `npx nx run compiler:build|test|typecheck`.
- From package: `npm run build`, `npm run dev`, `npm run test`, `npm run coverage`, `npm run typecheck`.
- Artifacts in `dist/` must exist before root Vite build when APIs change.

## Coding Style
- TypeScript (strict). ESLint + `@typescript-eslint` and `import/order`.
- Prettier: tabs, single quotes, semi, width 120, trailing commas.

## Testing
- Jest with `@swc/jest`. Place tests in `tests/`, `__tests__/`, or `*.test.ts`.
- Focus on deterministic, fast unit tests for parsing, IR, and transforms.

## Commits & PRs
- Commits: `compiler: <scope> <change>` (e.g., `compiler: parser fix for arrays`).
- PRs: include rationale, test coverage notes, and linked issues.
