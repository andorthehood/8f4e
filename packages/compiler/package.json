{
	"name": "@8f4e/compiler",
	"version": "1.0.0",
	"description": "8f4e compiler package",
	"type": "module",
	"main": "dist/index.js",
	"types": "dist/index.d.ts",
	"exports": {
		".": {
			"types": "./dist/index.d.ts",
			"default": "./dist/index.js"
		}
	},
	"sideEffects": false,
	"scripts": {
		"test": "jest tests",
		"coverage": "jest --coverage",
		"build": "tsc",
		"dev": "tsc --watch",
		"typecheck": "tsc --noEmit"
	},
	"repository": {
		"type": "git",
		"url": "git+https://github.com/andorthehood/8f4e.git"
	},
	"author": "Andor Polgar",
	"license": "MIT",
	"bugs": {
		"url": "https://github.com/andorthehood/8f4e/issues"
	},
	"homepage": "https://github.com/andorthehood/8f4e#readme",
	"devDependencies": {
		"@types/node": "^22.18.0"
	}
}
