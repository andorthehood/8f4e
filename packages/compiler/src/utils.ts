import { GLOBAL_ALIGNMENT_BOUNDARY } from './consts';
import { BLOCK_TYPE } from './types';

import type { BlockStack, CompilationContext, MemoryMap, StackItem } from './types';

export function isMemoryIdentifier(memoryMap: MemoryMap, name: string): boolean {
	return name in memoryMap;
}

export function isMemoryReferenceIdentifier(memoryMap: MemoryMap, name: string): boolean {
	return (
		(name.startsWith('&') && name.substring(1) in memoryMap) || (name.endsWith('&') && name.slice(0, -1) in memoryMap)
	);
}

export function isMemoryPointerIdentifier(memoryMap: MemoryMap, name: string): boolean {
	return name.startsWith('*') && name.substring(1) in memoryMap;
}

export function isElementCountIdentifier(memoryMap: MemoryMap, name: string): boolean {
	return name.startsWith('$') && name.substring(1) in memoryMap;
}

export function isElementWordSizeIdentifier(memoryMap: MemoryMap, name: string): boolean {
	return name.startsWith('%') && name.substring(1) in memoryMap;
}

export function getDataStructure(memoryMap: MemoryMap, id: string) {
	return memoryMap[id];
}

export function getDataStructureByteAddress(memoryMap: MemoryMap, id: string): number {
	const memoryItem = getDataStructure(memoryMap, id);
	return memoryItem ? memoryItem.byteAddress : 0;
}

export function getMemoryStringLastByteAddress(memoryMap: MemoryMap, id: string): number {
	const memoryItem = getDataStructure(memoryMap, id);
	return memoryItem ? memoryItem.byteAddress + (memoryItem.wordAlignedSize - 1) * GLOBAL_ALIGNMENT_BOUNDARY : 0;
}

export function getElementCount(memoryMap: MemoryMap, id: string): number {
	const memoryItem = getDataStructure(memoryMap, id);
	return memoryItem ? memoryItem.numberOfElements : 0;
}

export function getElementWordSize(memoryMap: MemoryMap, id: string): number {
	const memoryItem = getDataStructure(memoryMap, id);
	return memoryItem ? memoryItem.elementWordSize : 0;
}

export function isInstructionIsInsideAModule(blockStack: BlockStack) {
	for (let i = blockStack.length - 1; i >= 0; i--) {
		if (blockStack[i].blockType === BLOCK_TYPE.MODULE) {
			return true;
		}
	}
	return false;
}

export function isInstructionIsInsideBlock(blockStack: BlockStack, blockType: BLOCK_TYPE) {
	for (let i = blockStack.length - 1; i >= 0; i--) {
		if (blockStack[i].blockType === blockType) {
			return true;
		}
	}
	return false;
}

export function calculateWordAlignedSizeOfMemory(memory: MemoryMap): number {
	return Object.values(memory).reduce((accumulator, current) => {
		return accumulator + current.wordAlignedSize;
	}, 0);
}

export function areAllOperandsIntegers(...operands: StackItem[]): boolean {
	return !operands.some(operand => !operand.isInteger);
}

export function areAllOperandsFloats(...operands: StackItem[]): boolean {
	return !operands.some(operand => operand.isInteger);
}

export function saveByteCode(context: CompilationContext, byteCode: number[]): CompilationContext {
	if (isInstructionIsInsideBlock(context.blockStack, BLOCK_TYPE.INIT)) {
		context.initSegmentByteCode.push(...byteCode);
	} else {
		context.loopSegmentByteCode.push(...byteCode);
	}
	return context;
}
