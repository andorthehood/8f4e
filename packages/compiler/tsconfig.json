{
	"extends": "../../tsconfig.json",
	"compilerOptions": {
		"outDir": "dist",
		"rootDir": "src",
		"noEmit": false,
		"declaration": true,
		"declarationMap": true,
		"sourceMap": true,
		"types": []
	},
	"include": [
		"src/**/*.ts"
	],
	"exclude": [
		"node_modules",
		"src/**/*.test.ts"
	]
}
