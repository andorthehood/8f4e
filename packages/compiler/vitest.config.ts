import { defineConfig, mergeConfig } from 'vitest/config';

import { vitestPreset } from '../../vitest.preset';

export default mergeConfig(
	vitestPreset,
	defineConfig({
		test: {
			include: ['tests/**/*.test.ts'],
			exclude: ['**/node_modules/**', '**/dist/**', '**/testUtils.ts', '**/__fixtures__/**'],
		},
	})
);
