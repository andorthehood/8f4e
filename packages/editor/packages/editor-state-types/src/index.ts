// Re-export types needed by web-ui and other consumers
export type { State, CodeBlockGraphicData } from './types';

// Also export types that editor consumers need
export type {
	Project,
	Options,
	EditorSettings,
	CompilationResult,
	CodeBlock,
	Viewport,
	Size,
	Position,
	ContextMenuItem,
	MenuGenerator,
	BuildError,
	Compiler,
	Midi,
	GraphicHelper,
	BinaryAsset,
	Callbacks,
	ExampleModule,
	ModuleMetadata,
	ProjectMetadata,
	Runtimes,
	WebWorkerLogicRuntime,
	MainThreadLogicRuntime,
	AudioWorkletRuntime,
	WebWorkerMIDIRuntime,
	MemoryIdentifier,
	BufferPlotter,
	Switch,
	Debugger,
	Output,
	Input,
	PianoKeyboard,
	ContextMenu,
} from './types';

// Export EMPTY_DEFAULT_PROJECT as a value
export { EMPTY_DEFAULT_PROJECT } from './types';

export type { RuntimeFactory, RuntimeType } from './runtime';
export type { FeatureFlags, FeatureFlagsConfig } from './featureFlags';
export type { EventDispatcher } from './eventDispatcher';
