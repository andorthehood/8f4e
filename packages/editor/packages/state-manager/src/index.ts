import type { Path, PathValue, Subscription } from './types';

export interface StateManager<State> {
	getState: () => State;
	set: <P extends Path<State>>(selector: P, value: PathValue<State, P>) => void;
	subscribe: <P extends Path<State>>(
		selector: P,
		callback: (value: PathValue<State, P>) => void
	) => Subscription<State, P>;
	unsubscribe: <P extends Path<State>>(subscription: Subscription<State, P>) => void;
}

function createStateManager<State>(state: State): StateManager<State> {
	const subscriptions = new Set<Subscription<State>>();

	return {
		getState: () => state,
		set<P extends Path<State>>(selector: P, value: PathValue<State, P>) {
			const path = String(selector);
			const keys = path.split('.');
			// eslint-disable-next-line @typescript-eslint/no-explicit-any
			let current: any = state;

			if (keys.length === 1) {
				current[keys[0]] = value;
			} else {
				for (let i = 0; i < keys.length - 1; i++) {
					current = current[keys[i]];
				}
				current[keys[keys.length - 1]] = value;
			}

			subscriptions.forEach(subscription => {
				const { tokens, callback } = subscription;
				const comparisonLength = Math.min(tokens.length, keys.length);

				for (let i = 0; i < comparisonLength; i++) {
					if (tokens[i] !== keys[i]) {
						return;
					}
				}

				let target: unknown = state;
				for (const token of tokens) {
					if (target === null) {
						target = undefined;
						break;
					}

					if (typeof target === 'object' || typeof target === 'function') {
						target = (target as Record<string, unknown>)[token];
					} else {
						target = undefined;
						break;
					}
				}

				(callback as (value: unknown) => void)(target);
			});
		},
		subscribe<P extends Path<State>>(selector: P, callback: (value: PathValue<State, P>) => void) {
			const tokens = String(selector).split('.');
			// eslint-disable-next-line @typescript-eslint/no-explicit-any
			const subscription: Subscription<State, P> = {
				selector,
				tokens,
				callback,
			};
			subscriptions.add(subscription as unknown as Subscription<State>);
			return subscription;
		},
		unsubscribe<P extends Path<State>>(subscription: Subscription<State, P>) {
			subscriptions.delete(subscription as unknown as Subscription<State>);
		},
	};
}

export default createStateManager;
