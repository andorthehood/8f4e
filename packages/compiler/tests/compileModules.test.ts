import modules from './__fixtures__/modules';

import { compileToAST } from '../src/compiler';
import { compileModules } from '../src';
import compile from '../src';

describe('compiler', () => {
	test('compileModules', () => {
		const astModules = modules.map(({ code }) => compileToAST(code));
		expect(
			compileModules(astModules, {
				startingMemoryWordAddress: 0,
				environmentExtensions: { constants: {}, ignoredKeywords: [] },
				maxMemorySize: 1,
				initialMemorySize: 1,
				includeAST: true,
			})
		).toMatchSnapshot();
	});

	test('compileModules excludes AST when includeAST is false', () => {
		const astModules = modules.map(({ code }) => compileToAST(code));
		const compiledModules = compileModules(astModules, {
			startingMemoryWordAddress: 0,
			environmentExtensions: { constants: {}, ignoredKeywords: [] },
			maxMemorySize: 1,
			initialMemorySize: 1,
			includeAST: false,
		});

		// Verify that none of the compiled modules have an ast property
		compiledModules.forEach(module => {
			expect(module.ast).toBeUndefined();
		});
	});

	test('compile function excludes AST by default', () => {
		const result = compile(modules, {
			startingMemoryWordAddress: 0,
			environmentExtensions: { constants: {}, ignoredKeywords: [] },
			maxMemorySize: 1,
			initialMemorySize: 1,
		});

		// Verify that none of the compiled modules have an ast property
		for (const [, module] of result.compiledModules) {
			expect(module.ast).toBeUndefined();
		}
	});

	test('compile function includes AST when includeAST is true', () => {
		const result = compile(modules, {
			startingMemoryWordAddress: 0,
			environmentExtensions: { constants: {}, ignoredKeywords: [] },
			maxMemorySize: 1,
			initialMemorySize: 1,
			includeAST: true,
		});

		// Verify that all compiled modules have an ast property
		for (const [, module] of result.compiledModules) {
			expect(module.ast).toBeDefined();
			expect(Array.isArray(module.ast)).toBe(true);
		}
	});
});
