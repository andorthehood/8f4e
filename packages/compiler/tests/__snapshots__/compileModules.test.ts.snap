// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compiler compileModules 1`] = `
[
  {
    "ast": [
      {
        "arguments": [
          {
            "type": "identifier",
            "value": "abs",
          },
        ],
        "instruction": "module",
        "lineNumber": 1,
      },
      {
        "arguments": [
          {
            "type": "identifier",
            "value": "DEFAULT_VALUE",
          },
          {
            "isInteger": true,
            "type": "literal",
            "value": 0,
          },
        ],
        "instruction": "int",
        "lineNumber": 4,
      },
      {
        "arguments": [
          {
            "type": "identifier",
            "value": "in",
          },
          {
            "type": "identifier",
            "value": "&DEFAULT_VALUE",
          },
        ],
        "instruction": "int*",
        "lineNumber": 5,
      },
      {
        "arguments": [
          {
            "type": "identifier",
            "value": "out",
          },
          {
            "isInteger": true,
            "type": "literal",
            "value": 0,
          },
        ],
        "instruction": "int",
        "lineNumber": 6,
      },
      {
        "arguments": [
          {
            "type": "identifier",
            "value": "test",
          },
          {
            "isInteger": true,
            "type": "literal",
            "value": 1,
          },
        ],
        "instruction": "int",
        "lineNumber": 7,
      },
      {
        "arguments": [
          {
            "type": "identifier",
            "value": "int",
          },
          {
            "type": "identifier",
            "value": "input",
          },
        ],
        "instruction": "local",
        "lineNumber": 10,
      },
      {
        "arguments": [
          {
            "type": "identifier",
            "value": "&out",
          },
        ],
        "instruction": "push",
        "lineNumber": 13,
      },
      {
        "arguments": [
          {
            "type": "identifier",
            "value": "in",
          },
        ],
        "instruction": "push",
        "lineNumber": 14,
      },
      {
        "arguments": [
          {
            "type": "identifier",
            "value": "input",
          },
        ],
        "instruction": "localSet",
        "lineNumber": 15,
      },
      {
        "arguments": [
          {
            "type": "identifier",
            "value": "input",
          },
        ],
        "instruction": "push",
        "lineNumber": 16,
      },
      {
        "arguments": [
          {
            "isInteger": true,
            "type": "literal",
            "value": 0,
          },
        ],
        "instruction": "push",
        "lineNumber": 17,
      },
      {
        "arguments": [],
        "instruction": "lessThan",
        "lineNumber": 18,
      },
      {
        "arguments": [],
        "instruction": "if",
        "lineNumber": 19,
      },
      {
        "arguments": [
          {
            "isInteger": true,
            "type": "literal",
            "value": 0,
          },
        ],
        "instruction": "push",
        "lineNumber": 20,
      },
      {
        "arguments": [
          {
            "type": "identifier",
            "value": "input",
          },
        ],
        "instruction": "push",
        "lineNumber": 21,
      },
      {
        "arguments": [],
        "instruction": "sub",
        "lineNumber": 22,
      },
      {
        "arguments": [],
        "instruction": "else",
        "lineNumber": 23,
      },
      {
        "arguments": [
          {
            "type": "identifier",
            "value": "input",
          },
        ],
        "instruction": "push",
        "lineNumber": 24,
      },
      {
        "arguments": [],
        "instruction": "ifEnd",
        "lineNumber": 25,
      },
      {
        "arguments": [],
        "instruction": "store",
        "lineNumber": 26,
      },
      {
        "arguments": [],
        "instruction": "moduleEnd",
        "lineNumber": 27,
      },
    ],
    "byteAddress": 0,
    "id": "abs",
    "index": 0,
    "initFunctionBody": [],
    "loopFunction": [
      32,
      1,
      1,
      127,
      65,
      8,
      65,
      4,
      40,
      2,
      0,
      33,
      0,
      32,
      0,
      65,
      0,
      72,
      4,
      127,
      65,
      0,
      32,
      0,
      107,
      5,
      32,
      0,
      11,
      54,
      2,
      0,
      11,
    ],
    "memoryMap": {
      "DEFAULT_VALUE": {
        "byteAddress": 0,
        "default": 0,
        "elementWordSize": 4,
        "id": "DEFAULT_VALUE",
        "isInteger": true,
        "isPointer": false,
        "isPointingToInteger": false,
        "isPointingToPointer": false,
        "numberOfElements": 1,
        "type": "int",
        "wordAlignedAddress": 0,
        "wordAlignedSize": 1,
      },
      "in": {
        "byteAddress": 4,
        "default": 0,
        "elementWordSize": 4,
        "id": "in",
        "isInteger": true,
        "isPointer": true,
        "isPointingToInteger": true,
        "isPointingToPointer": false,
        "numberOfElements": 1,
        "type": "int*",
        "wordAlignedAddress": 1,
        "wordAlignedSize": 1,
      },
      "out": {
        "byteAddress": 8,
        "default": 0,
        "elementWordSize": 4,
        "id": "out",
        "isInteger": true,
        "isPointer": false,
        "isPointingToInteger": false,
        "isPointingToPointer": false,
        "numberOfElements": 1,
        "type": "int",
        "wordAlignedAddress": 2,
        "wordAlignedSize": 1,
      },
      "test": {
        "byteAddress": 12,
        "default": 1,
        "elementWordSize": 4,
        "id": "test",
        "isInteger": true,
        "isPointer": false,
        "isPointingToInteger": false,
        "isPointingToPointer": false,
        "numberOfElements": 1,
        "type": "int",
        "wordAlignedAddress": 3,
        "wordAlignedSize": 1,
      },
    },
    "wordAlignedAddress": 0,
    "wordAlignedSize": 4,
  },
]
`;
